// @nestjs/microservices/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TCP_DEFAULT_PORT = 3000;
exports.TCP_DEFAULT_HOST = 'localhost';
exports.REDIS_DEFAULT_URL = 'redis://localhost:6379';
exports.NATS_DEFAULT_URL = 'nats://localhost:4222';
exports.MQTT_DEFAULT_URL = 'mqtt://localhost:1883';
exports.GRPC_DEFAULT_URL = 'localhost:5000';
exports.RQM_DEFAULT_URL = 'amqp://localhost';
exports.CONNECT_EVENT = 'connect';
exports.DISCONNECT_EVENT = 'disconnect';
exports.MESSAGE_EVENT = 'message';
exports.DATA_EVENT = 'data';
exports.ERROR_EVENT = 'error';
exports.CLOSE_EVENT = 'close';
exports.SUBSCRIBE = 'subscribe';
exports.CANCEL_EVENT = 'cancelled';
exports.PATTERN_METADATA = 'microservices:pattern';
exports.CLIENT_CONFIGURATION_METADATA = 'microservices:client';
exports.PATTERN_HANDLER_METADATA = 'microservices:handler_type';
exports.CLIENT_METADATA = 'microservices:is_client_instance';
exports.RQM_DEFAULT_QUEUE = 'default';
exports.RQM_DEFAULT_PREFETCH_COUNT = 0;
exports.RQM_DEFAULT_IS_GLOBAL_PREFETCH_COUNT = false;
exports.RQM_DEFAULT_QUEUE_OPTIONS = {};
exports.GRPC_DEFAULT_PROTO_LOADER = '@grpc/proto-loader';
exports.NO_MESSAGE_HANDLER = `There is no matching message handler defined in the remote service.`;
exports.NO_EVENT_HANDLER = `There is no matching event handler defined in the remote service.`;
exports.DISCONNECTED_RMQ_MESSAGE = `Disconnected from RMQ. Trying to reconnect.`;
exports.GRPC_DEFAULT_MAX_RECEIVE_MESSAGE_LENGTH = 4 * 1024 * 1024;
exports.GRPC_DEFAULT_MAX_SEND_MESSAGE_LENGTH = 4 * 1024 * 1024;
